crates/noirc_evaluator/src/ssa_refactor/acir_gen/mod.rs

```diff
@@ -31,21 +34,28 @@ struct Context {
     /// for an SSA value, we check this map. If an `AcirVar`
     /// already exists for this Value, we return the `AcirVar`.
     ssa_value_to_acir_var: HashMap<Id<Value>, AcirVar>,
+    /// Maps SSA values to array addresses (including index offset).
     ///
+    /// When converting parameters the of main, `ArrayId`s are gathered and stored with an offset
+    /// of 0. When the use of these stored values are detected for address arithmetic, the results
+    /// of such instructions are stored, in effect capturing any further values that refer to
+    /// addresses.
+    ssa_value_to_array_address: HashMap<ValueId, (ArrayId, usize)>,
+    /// Manages and builds the `AcirVar`s to which the converted SSA values refer.
     acir_context: AcirContext,
 }
 
 impl Ssa {
     pub(crate) fn into_acir(self, main_function_signature: FunctionSignature) -> GeneratedAcir {
-        let _param_array_lengths = collate_array_lengths(&main_function_signature.0);
+        let param_array_lengths = collate_array_lengths(&main_function_signature.0);
```

Comment:
> Or maybe we should just look at adding lengths for references in SSA.

Sounds reasonable to me - at the same time we should also add `NumericType` of the elements. See this PR for why we need it: https://github.com/noir-lang/noir/pull/1460/

Reply:
I'd be a bit more worried about adding NumericType elements versus lengths since it may make supporting nested arrays more difficult later. It'd necessitate either removing that addition or trying to flatten nested arrays into a single array.

Reply:
One more thing, can you elaborate on why we use the types of the ssa parameters to translate main's parameters in Acir rather than the function AbiTypes? Originally it was intended to use the later, hence why References in SSA in particular have so little information. Was it because of the array of structs transformation of monomorphization? If so perhaps we consider adding an `if experimental_ssa { ... }` check in monomorphization to remove it since it seems like it is complicating things.

Reply:
> rather than the function AbiTypes

Mainly I was piggy-backing off Kev's `numeric_cast_var` impl, which accepted that type. It's probably ok to modify it though.

Removing the monomorphization struct transform may help to decomplicate things, but I think doing so might be worthy of a separate issue and PR. (We already have [this one](https://github.com/noir-lang/noir/issues/1425).) 

Would it feel more intuitive if inverted the consumption relationship? I.e. Perform the various `ValueId`/`AcirVar`/`ArrayId` bindings while depth-first traversing the abi (consuming block params on demand via `block_params.next()`); as opposed to stepping through the block params and consuming abi info via `flattened_abi.next()`?

Reply:
I think recursing with the current `block_params.next()` makes more sense to me than the reverse with `flattened_abi.next()`. Lets just keep this as-is and think about some solutions to clean it up later. Like adding lengths to the reference type. I think the lengths may be needed anyway in order to implement array copying.